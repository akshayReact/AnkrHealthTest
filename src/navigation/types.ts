import type { Paths } from '@/navigation/paths';
import type { StackScreenProps } from '@react-navigation/stack';

export type RootScreenProps<
  S extends keyof RootStackParamList = keyof RootStackParamList,
> = StackScreenProps<RootStackParamList, S>;

export type RootStackParamList = {
  [Paths.Example]: undefined;
  [Paths.Startup]: undefined;
};

export interface JobPosting {
  id: string;
  title: string;
  description: string;
  company: string;
  location: string;
  salary_from: number;
  salary_to: number;
  employment_type: string;
  application_deadline: string; // You can use Date if you parse it
  qualifications: string; // JSON string; you may also parse this to `string[]`
  contact: string;
  job_category: string;
  is_remote_work: number; // Use boolean if you map 0/1 to false/true
  number_of_opening: number;
  created_at: string;
  updated_at: string;
}

export interface FilterJobType {
  label: string,
  value: string
}

export interface JobFilter {
  page: number,
  jobType?: string,
  search?: string
}
