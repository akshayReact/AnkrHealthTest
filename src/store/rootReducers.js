import home from '@/apis/home/homeSlice'
export const rootReducer = {
  home
}

export type RootState = ReturnType<typeof rootReducer>
