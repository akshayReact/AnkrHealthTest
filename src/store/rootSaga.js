import { all } from 'redux-saga/effects'
import homeSaga from '@/apis/home/homeApis'
// Here you can include all the saga which you write for components

export default function* rootSaga() {
  // yield all([AuthenticationSaga()]),
  // yield all([FormsSaga()])
  yield all([homeSaga()]);
}
