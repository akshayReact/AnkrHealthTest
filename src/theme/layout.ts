import { Platform, type ViewStyle } from 'react-native';

export default {
  col: {
    flexDirection: 'column',
  },
  colReverse: {
    flexDirection: 'column-reverse',
  },
  itemsCenter: {
    alignItems: 'center',
  },
  itemsEnd: {
    alignItems: 'flex-end',
  },
  itemsStart: {
    alignItems: 'flex-start',
  },
  itemsStretch: {
    alignItems: 'stretch',
  },
  justifyAround: {
    justifyContent: 'space-around',
  },
  justifyBetween: {
    justifyContent: 'space-between',
  },
  justifyCenter: {
    justifyContent: 'center',
  },
  justifyEnd: {
    justifyContent: 'flex-end',
  },
  justifyStart: {
    justifyContent: 'flex-start',
  },
  row: {
    flexDirection: 'row',
  },
  rowReverse: {
    flexDirection: 'row-reverse',
  },
  wrap: {
    flexWrap: 'wrap',
  },
  /* Sizes Layouts */
  flex_1: {
    flex: 1,
  },
  fullHeight: {
    height: '100%',
  },
  fullWidth: {
    width: '100%',
  },
  /* Positions */
  absolute: {
    position: 'absolute',
  },
  bottom0: {
    bottom: 0,
  },
  left0: {
    left: 0,
  },
  relative: {
    position: 'relative',
  },
  right0: {
    right: 0,
  },
  top0: {
    top: 0,
  },
  z1: {
    zIndex: 1,
  },
  z10: {
    zIndex: 10,
  },
  inputFieldStyle: {
    zIndex: 100,
    borderColor: 'rgba(0,0,0,0.1)',
    paddingHorizontal: 10,
    backgroundColor: 'rgba(255, 255, 255, 1)',
    borderRadius: 10,
    borderWidth: 1
  },
  textBox1: {
    height: 50,
    paddingLeft: 10,
  },
} as const satisfies Record<string, ViewStyle>;
