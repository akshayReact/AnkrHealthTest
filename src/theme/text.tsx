import { TextStyle } from "react-native";

export default {
  rowItemTitle: {
    fontSize: 14,
    color: "#585856",
    fontFamily: "Poppins-regular"
  },
} as const satisfies Record<string, TextStyle>;
