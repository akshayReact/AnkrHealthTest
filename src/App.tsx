import 'react-native-gesture-handler';

import { QueryClient, QueryClientProvider } from '@tanstack/react-query';
import { GestureHandlerRootView } from 'react-native-gesture-handler';
import { MMKV } from 'react-native-mmkv';
import { Provider as ReduxProvider } from "react-redux";
import store from "@/store/store";
import ApplicationNavigator from '@/navigation/Application';
import { ThemeProvider } from '@/theme';
import '@/translations';

export const queryClient = new QueryClient({
  defaultOptions: {
    mutations: {
      retry: false,
    },
    queries: {
      retry: false,
    },
  },
});

export const storage = new MMKV();

function App() {
  return (
    <GestureHandlerRootView>
      <ReduxProvider store={store}>
      <QueryClientProvider client={queryClient}>
        <ThemeProvider storage={storage}>
          <ApplicationNavigator />
        </ThemeProvider>
      </QueryClientProvider>
      </ReduxProvider>
    </GestureHandlerRootView>
  );
}

export default App;
