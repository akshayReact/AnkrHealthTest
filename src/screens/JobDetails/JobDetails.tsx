import { useTranslation } from 'react-i18next';
import { Alert, StyleSheet, Text, TouchableOpacity, View } from 'react-native';
import { useTheme } from '@/theme';
import { IconByVariant, SubmitButton } from '@/components/atoms';
import { SafeScreen } from '@/components/templates';
import { useDispatch, useSelector } from 'react-redux';
import { RootState } from '@/store/rootReducers';

function JobDetails({ navigation }) {
    const { t } = useTranslation();
    const dispatch = useDispatch();
    const { selectedJob } = useSelector(
        (state: RootState) => state.home
    );
    const {
        colors,
        components,
        fonts,
        gutters,
        layout,
    } = useTheme();

    const handleBtnPress = async () => {
        await Alert.alert("Applied Successfully!");
        handleGoBack();
    }

    const handleGoBack = () => {
        navigation.goBack()
    }

    return (
        <SafeScreen
            isError={false}
        >
            <View
                style={[
                    layout.row,
                    layout.justifyBetween,
                    layout.itemsCenter,
                    gutters.marginTop_16,
                    gutters.margin_12,
                ]}
            >
                <TouchableOpacity
                    onPress={() => handleGoBack()}
                    style={[components.buttonCircle]}
                    testID="back-button"
                >
                    <IconByVariant path="leftarrow" stroke={colors.purple500} height={18} width={18} />
                </TouchableOpacity>
                <Text style={[fonts.size_16, fonts.gray800, fonts.bold]}>
                    Job Details
                </Text>
                <View></View>
            </View>
            <View style={[gutters.marginTop_40, gutters.marginBottom_16]}>
                <Text testID='selected-job-title' numberOfLines={2} style={[fonts.size_24, fonts.gray800, fonts.bold, gutters.marginLeft_12, {flexWrap: 'wrap'}]}>
                    {selectedJob?.title}
                </Text>
            </View>
            <View style={styles.card} testID='job-details-container'>
                <Text testID='job-title' style={styles.title}>{selectedJob?.title}</Text>
                <Text testID='job-company-name' style={styles.company}>{selectedJob?.company}</Text>
                <Text testID='job-location' style={styles.label}>Location: <Text style={styles.value}>{selectedJob?.location}</Text></Text>
                <Text testID='job-category' style={styles.label}>Job Category: <Text style={styles.value}>{selectedJob?.job_category}</Text></Text>
                <Text testID='job-employment-type' style={styles.label}>Employment Type: <Text style={styles.value}>{selectedJob?.employment_type}</Text></Text>
                <Text testID='job-is-remote' style={styles.label}>Remote: <Text style={styles.value}>{selectedJob?.is_remote_work ? 'Yes' : 'No'}</Text></Text>
                <Text testID='job-openings' style={styles.label}>Openings: <Text style={styles.value}>{selectedJob?.number_of_opening}</Text></Text>
                <Text testID='job-salary' style={styles.label}>Salary: <Text style={styles.value}>${selectedJob?.salary_from} - ${selectedJob?.salary_to}</Text></Text>
                <Text testID='job-deadline' style={styles.label}>Deadline: <Text style={styles.value}>{selectedJob?.application_deadline}</Text></Text>
                <Text testID='job-contact' style={styles.label}>Contact: <Text style={styles.value}>{selectedJob?.contact}</Text></Text>
                <Text testID='job-qualifications' style={styles.label}>Qualifications:</Text>
                {JSON.parse(selectedJob?.qualifications || '[]').map((q, index) => (
                    <Text key={index} style={styles.qualification}>â€¢ {q}</Text>
                ))}
            </View>
            <SubmitButton onPress={handleBtnPress} testID='submit-btn' />
        </SafeScreen>
    );
}

export default JobDetails;

const styles = StyleSheet.create({
    card: {
        backgroundColor: '#fff',
        padding: 15,
        marginVertical: 8,
        marginHorizontal: 12,
        borderRadius: 8,
        elevation: 2
    },
    title: {
        fontSize: 18,
        fontWeight: 'bold'
    },
    company: {
        fontSize: 16,
        color: '#555',
        marginBottom: 6
    },
    label: {
        fontWeight: '600',
        marginTop: 4
    },
    value: {
        fontWeight: '400'
    },
    qualification: {
        marginLeft: 10,
        fontStyle: 'italic'
    }
});