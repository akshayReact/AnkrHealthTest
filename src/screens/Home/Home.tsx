import { useEffect, useState } from 'react';
import { FlatList, Text, ActivityIndicator, View, StyleSheet, TouchableOpacity, TouchableHighlight } from 'react-native';

import { useTheme } from '@/theme';

import { SafeScreen } from '@/components/templates';
import { useDispatch, useSelector } from 'react-redux';
import { getJobsList, setSelectedJobData } from '@/apis/home/homeSlice';
import { RootState } from '@/store/rootReducers';
import JobItemCard from './Components/JobItemCard';
import text from '@/theme/text';
import { JobFilter, JobPosting } from '@/navigation/types';
import { CommonActions } from '@react-navigation/native';
import { FILTER_JOB_TYPE, FILTER_JOB_TYPE_OPTIONS } from '@/Constants';
import TextBox from '@/components/atoms/TextBox/TextBox';
import DropDownPicker from 'react-native-dropdown-picker';

const MAX_RANDOM_ID = 9;
let navigateTimer: any = null;
let loadMoreTimer: any = null;

function Home({ navigation }) {
  const { homeJobs, isLoading, nextPageUrl } = useSelector(
    (state: RootState) => state.home
  );
  const dispatch = useDispatch();
  const [jobList, setJobList] = useState(homeJobs)
  const [filters, setFilters] = useState<JobFilter>({
    page: 1
  });
  const [isEndReached, setIsEndReached] = useState(false);
  const [searchText, setSearchText] = useState('');
  const [open, setOpen] = useState(false);
  const [value, setValue] = useState(null);
  const [items, setItems] = useState(FILTER_JOB_TYPE_OPTIONS)
  const {
    fonts,
    gutters,
    layout,
  } = useTheme();

  useEffect(() => {
    dispatch(getJobsList(filters))
  }, [filters])

  useEffect(() => {
    setJobList(homeJobs)
  }, [homeJobs])

  useEffect(() => {
    if (value) {
      onSelectFilter(value);
    }
  }, [value])

  const handleEndReached = () => {
    if (jobList.length > 10) {
      clearTimeout(loadMoreTimer);
      setIsEndReached(true)
      loadMoreTimer = setTimeout(() => {
        setFilters(prevState => {
          return {
            ...prevState,
            page: prevState.page + 1
          }
        });
        setIsEndReached(false);
      }, 1500)
    }
  };

  const handleJobItemPress = async (jobItem: JobPosting) => {
    clearTimeout(navigateTimer)
    await dispatch(setSelectedJobData(jobItem))
    navigateTimer = setTimeout(() => {
      navigation.dispatch(
        CommonActions.navigate({
          name: "jobDetails",
        })
      );
    }, 500)
  }

  const onSelectFilter = (filterValue: string) => {
    try {
      setFilters(prevState => {
        return {
          ...prevState,
          page: 1,
          jobType: filterValue
        }
      });
    } catch (err) {
      console.error(err)
    }
  }

  const onRemoveFilters = () => {
    setFilters({ page: 1 });
    setOpen(false);
    setValue(null);
    setItems(FILTER_JOB_TYPE);
  }

  const handleSearchTextChange = (txt: string) => {
    setSearchText(txt)
    const jobsPayload = homeJobs.filter(job =>
      job.company.toLocaleLowerCase().includes(txt.toLocaleLowerCase())
    );
    setJobList(jobsPayload)
  }

  return (
    <SafeScreen>
      <View style={[gutters.marginTop_40, gutters.marginBottom_16]}>
        <Text style={[fonts.size_24, fonts.gray800, fonts.bold, gutters.marginLeft_12]}>
          Anker Job Board
        </Text>
      </View>
      <View style={[layout.row, layout.itemsCenter, layout.justifyCenter, gutters.marginHorizontal_12, gutters.marginBottom_16]}>
        <TextBox
          hintText="Search by Company name"
          inputValue={searchText}
          handleInputChange={handleSearchTextChange}
        />
      </View>
      <View style={[layout.row, layout.itemsCenter, layout.justifyBetween, gutters.marginHorizontal_12, gutters.marginBottom_16]}>
        <Text style={[fonts.size_12, fonts.gray800]}>
          Filter By
        </Text>
        <DropDownPicker
          testID='job-filter-dropdown'
          open={open}
          style={[
            layout.inputFieldStyle,
            gutters.marginLeft_12,
            {
              height: 35,
              width: 240
            },
          ]}
          value={value}
          items={items}
          setOpen={setOpen}
          setValue={setValue}
          setItems={setItems}
        />
      </View>
      <Text testID='selected-filter' style={{fontSize: 16, color: '#000'}}>{value}</Text>

      {
        filters.jobType ?
          <TouchableOpacity onPress={onRemoveFilters} testID="remove-filters">
            <Text style={styles.clickableText}>Remove Filters</Text>
          </TouchableOpacity>
          : null
      }
      {
        !jobList && isLoading ?
          <View style={[layout.flex_1, layout.itemsCenter, layout.justifyCenter]}>
            <ActivityIndicator
              size="large"
              color="#585856"
            />
            <Text
              style={[fonts.size_16, fonts.gray200, gutters.marginTop_12]}
            >
              Loading Jobs. Please wait..
            </Text>
          </View>
          : ((!jobList && !isLoading) || (Array.isArray(jobList) && jobList.length == 0 && !isLoading)) ?
            <View style={[layout.flex_1, layout.itemsCenter, layout.justifyCenter]}>
              <Text
                style={[fonts.size_16, fonts.gray200, gutters.marginBottom_40]}
              >
                Nothing to show
              </Text>
            </View>
            :
            jobList && Array.isArray(jobList) && !isLoading ?
              <FlatList
                data={jobList}
                renderItem={({ item, index }) => (
                  <TouchableHighlight onPress={() => handleJobItemPress(item)} testID={`row-${(index + 1)}`}>
                    <JobItemCard jobItem={item} />
                  </TouchableHighlight>
                )}
                keyExtractor={(item) => {
                  return `${Math.random() * 10} ${item.id}`
                }}
                keyboardShouldPersistTaps="handled"
                showsVerticalScrollIndicator={false}
                onEndReached={({ distanceFromEnd }) => {
                  handleEndReached();
                }}
                onEndReachedThreshold={0.9} // Trigger when 50% of the list is scrolled
                initialNumToRender={10}
                ListFooterComponent={
                  <View
                    style={[
                      layout.row,
                      gutters.padding_12,
                      layout.justifyCenter,
                      layout.itemsCenter,
                      layout.fullWidth,
                      {
                        height: 60
                      }
                    ]}
                  >
                    {
                      isEndReached && nextPageUrl ?
                        <ActivityIndicator
                          size="small"
                          color="#585856"
                        />
                        :
                        isEndReached && !nextPageUrl ?
                          <Text
                            style={[
                              text.rowItemTitle,
                              {
                                color: "rgba(100, 100, 100, 1)",
                                fontWeight: 300,
                                fontSize: 16,
                              },
                            ]}
                          >
                            End of Results
                          </Text>
                          :
                          null
                    }
                  </View>
                }
              />
              :
              null
      }
    </SafeScreen>
  );
}

export default Home;

const styles = StyleSheet.create({
  dropdownText: { paddingLeft: 10, fontSize: 12 },
  clickableText: {
    color: '#1e90ff',
    fontSize: 16,
    textDecorationLine: 'underline',
    margin: 10
  }
})