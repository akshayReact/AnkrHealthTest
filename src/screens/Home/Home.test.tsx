import { fireEvent, render, screen, waitFor } from '@testing-library/react-native';
import Home from './Home';
import { Provider } from 'react-redux';
import store from "@/store/store";
import { queryClient } from '@/App';
import { ThemeProvider } from '@/theme';
import { QueryClient, QueryClientProvider } from '@tanstack/react-query';
import { MMKV } from 'react-native-mmkv';
import { SafeAreaProvider } from 'react-native-safe-area-context';
import { NavigationContainer } from '@react-navigation/native';
import { createStackNavigator } from '@react-navigation/stack';
import { RootStackParamList } from '@/navigation/types';
import { Paths } from '@/navigation/paths';
import JobDetails from '../JobDetails/JobDetails';
import renderer from 'react-test-renderer';

const Stack = createStackNavigator<RootStackParamList>();

describe('Home screen should render correctly', () => {
  let storage: MMKV;
  const mockNav = { navigate: jest.fn() };
  const queryClient = new QueryClient({
    defaultOptions: {
      mutations: {
        gcTime: Infinity,
      },
      queries: {
        gcTime: Infinity,
        retry: false,
      },
    },
  });

  beforeAll(() => {
    storage = new MMKV();
  });
  //-------------- -------------- -------------- TEST 01 -------------- -------------- -------------- --------------
  test('Home page renders correctly', () => {
    const tree = renderer.create(<Home />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  //-------------- -------------- -------------- TEST 02 -------------- -------------- -------------- --------------

  test('the user can click on job to see details', async () => {
     const component = (
      <SafeAreaProvider>
        <Provider store={store}>
          <QueryClientProvider client={queryClient}>
            <ThemeProvider storage={storage}>
              <NavigationContainer>
                <Stack.Navigator screenOptions={{ headerShown: false }}>
                  <Stack.Screen component={Home} name={Paths.Home} />
                  <Stack.Screen component={JobDetails} name={Paths.JobDetails} />
                </Stack.Navigator>
              </NavigationContainer>
            </ThemeProvider>
          </QueryClientProvider>
        </Provider>
      </SafeAreaProvider>
    );
    const { findByTestId } = render(component);
    const jobItem = await findByTestId(`row-1`);

    fireEvent.press(jobItem);
    await waitFor(() => {
      expect(store.getState().home.selectedJob).toBeTruthy();
    });

  });

  //-------------- -------------- -------------- TEST 03 -------------- -------------- -------------- --------------

  test('should update the Search Field value and display data when user types', () => {
     const component = (
      <SafeAreaProvider>
        <Provider store={store}>
          <QueryClientProvider client={queryClient}>
            <ThemeProvider storage={storage}>
              <NavigationContainer>
                <Stack.Navigator screenOptions={{ headerShown: false }}>
                  <Stack.Screen component={Home} name={Paths.Home} />
                  <Stack.Screen component={JobDetails} name={Paths.JobDetails} />
                </Stack.Navigator>
              </NavigationContainer>
            </ThemeProvider>
          </QueryClientProvider>
        </Provider>
      </SafeAreaProvider>
    );
    const { getByTestId } = render(component);
    const textInput = getByTestId('search-input');
    fireEvent.changeText(textInput, 'Cumminns');
    // TextInput value should get updated
    expect(textInput.props.value).toBe('Cumminns');
  });

    //-------------- -------------- -------------- TEST 04 -------------- -------------- -------------- --------------

    test('Filter Dropdown renders correctly & shows list of options on press', async () => {
       const component = (
      <SafeAreaProvider>
        <Provider store={store}>
          <QueryClientProvider client={queryClient}>
            <ThemeProvider storage={storage}>
              <NavigationContainer>
                <Stack.Navigator screenOptions={{ headerShown: false }}>
                  <Stack.Screen component={Home} name={Paths.Home} />
                  <Stack.Screen component={JobDetails} name={Paths.JobDetails} />
                </Stack.Navigator>
              </NavigationContainer>
            </ThemeProvider>
          </QueryClientProvider>
        </Provider>
      </SafeAreaProvider>
    );
    const { getByTestId } = render(component);
    const dropdown = getByTestId('job-filter-dropdown');
    fireEvent.press(dropdown); // Simulate opening the dropdown
    expect(getByTestId('job-filter-dropdown').props.children[0][0].props.children[1].props.children).toBe('Select an item');
  });
});