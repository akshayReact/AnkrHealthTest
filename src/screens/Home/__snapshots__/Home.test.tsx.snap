// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home screen should render correctly Home page renders correctly 1`] = `null`;
